CONVENTIONS.md

```diff
@@ -546,3 +546,17 @@ used SCC there would be a delay before it could start. In recent versions of OCP
 (4.6+) this delay has been virtually eliminated, the usage of runlevels should
 not be required at all hence the primary alternative is to simply try the
 workload without any runlevel specified.
+
+#### Metrics
+
+Core operators deployed by the CVO as well as their operands are encouraged to expose metrics using the [Prometheus exposition format](https://prometheus.io/docs/instrumenting/exposition_formats/). Metrics are useful for alerting and telemetry. Operators can deploy [service monitors](https://docs.openshift.com/container-platform/latest/rest_api/monitoring_apis/servicemonitor-monitoring-coreos-com-v1.html) to have their metrics collected by the in-cluster Prometheus pods running in the `openshift-monitoring` namespace.
+
+When a core operator exposes Prometheus metrics, it must comply with the following requirements:
+
+1. It must use HTTPS. Metrics are confidential, particularly labels which can be identifiable and promising future state of "nothing confidential ever" is implausible.
+2. It must require authentication and it must support TLS client certificate authentication to avoid a dependency on the Kubernetes API server (see [Handling kube-apiserver disruption](#handling-kube-apiserver-disruption)). It may support bearer token authentication but the in-cluster Prometheus must be configured with TLS client certificate only (the certificate path to use is `/etc/prometheus/secrets/metrics-client-certs/tls.crt` and the key path is `/etc/prometheus/secrets/metrics-client-certs/tls.key`).
+3. It must support local authorization and always allow the well-know metrics scraping identity (`system:serviceaccount:openshift-monitoring:prometheus-k8s`) to access the /metrics endpoint. It may support delegated authorization check.
+
+The same requirements apply for operands managed by a core operator.
+
+Optional operators are strongly recommended to follow these requirements.
```

Comment:
Can we link to instructions or an example implementation or something, to help folks who are going to need to implement these requirements?

Reply:
Context is a lot.  

1. For library-go based operators and controllers, this happens with https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/pkg/cmd/operator/cmd.go#L12-L14
2. For controller-runtime operators and controllers, you can shim in a similar thing like networking operator here https://github.com/openshift/cluster-network-operator/blob/master/cmd/cluster-network-operator/main.go#L52
3. For code wiring their own chains, you can union the authorizers like this https://github.com/openshift/library-go/blob/47904dd9ff5adf225b83124c533177809fce28b4/pkg/controller/controllercmd/builder.go#L276-L281
4. for kube-rbac-proxy users, you can set the static authorization following this guide https://github.com/brancz/kube-rbac-proxy/blob/master/examples/static-auth/README.md and the operator must manage the --client-ca on-disk since I don't recall the in-cluster values being usable (ref https://github.com/brancz/kube-rbac-proxy/blob/master/pkg/authn/delegating.go#L62)

Reply:
Is it ok to link to https://rhobs-handbook.netlify.app/products/openshiftmonitoring/collecting_metrics.md/?
I feel that CONVENTIONS.md isn't the right place for going into the details. I'm also happy to move the rhobs-handbook content to https://github.com/openshift/enhancements/tree/master/dev-guide.

Reply:
@simonpasquier , if you have docs you're already maintaining with implementation tips and examples, then yes, definitely link to those. We don't need to duplicate that information, we just need to make sure people can find it. :-)